### ISSUE TITLE
Cursive support

### ISSUE THREAD

**Issue by ajoberstar** [clojurephant/clojurephant#66]
I haven't tried at all to see how this works, but it would be great to get integration with Cursive. Some of this may already work.

- [ ] Importing and understanding project structure (which source sets are clojure)
- [ ] Starting (and using) a REPL from your Gradle config

... anything else? ...

**Comment by ajoberstar** [comment]
@cmf Do you have any guidance on the starting point for this? Is this anything we can even handle from the plugin, or would it be purely a Cursive implementation issue?

**Comment by cursive-ide** [comment]
This is something that would have to be done in Cursive. I'll try to look at this soon, I think it's pretty easy.

**Comment by aadrian** [comment]
> ... anything else? ...

@ajoberstar 
1. having a `gradle-clojure` project template (or more) would be very useful.
Since Gradle does not support extensibility in the `init` task yet, maybe the Cursive plug-in can do this (or at least trigger the call), since IntelliJ (and some other IntelliJ plug-ins) do have various project templates.
2. "Intentions" (on the IntelliJ/Cursive side) to convert an existing Lein/Boot project to `gradle-clojure`. Gradle has the `init --type=pom` https://docs.gradle.org/current/userguide/build_init_plugin.html#sec:pom_maven_conversion_ that converts from an existing Maven project to a Gradle one. Something similar to convert from Lein/Boot to `gradle-clojure` would greatly help the adoption of this project.

IMHO both #1 and #2 are needed to increase the adoption of `gradle-clojure` and since this can't be done with Gradle alone (`init` tasks not extensible yet), the Cursive integration could help trigger it.

**Comment by ajoberstar** [comment]
@cursive-ide Thanks, Colin!

@aadrian I'd lean towards just dealing with those both in the plugin to keep it general to Gradle. Direct Cursive support on those should probably wait until those pieces are more stable. See the referenced issues for where I broke thouse out.

**Comment by ajoberstar** [comment]
@cursive-ide I should be able to put something together this weekend as a sample you can work from.

**Comment by cursive-ide** [comment]
@ajoberstar Actually I just realised that of course there are example projects etc in the README. I'll just use those, unless there's something different you'd like generated there. Once I have something working we can think about ways to coordinate the versions of gradle-clojure and Cursive, since I think the extensions that are generated into the build file are hard-coded.

**Comment by ajoberstar** [comment]
Yeah, I think starting with what's in the README is a good idea.

**Comment by cursive-ide** [comment]
I have a WIP, this is what it looks like now:

<img width="982" alt="screen shot 2018-05-29 at 10 35 11" src="https://user-images.githubusercontent.com/5589171/40631027-cc0fdb34-62c7-11e8-89db-9acd62dffc15.png">

Which generates this:

<img width="948" alt="screen shot 2018-05-29 at 10 36 52" src="https://user-images.githubusercontent.com/5589171/40631036-d7cfe86a-62c7-11e8-96bf-de5f0b2031f1.png">

The main problem with this is that the versions are all hard coded, and I can't see a good way to fix that. The first thing that occurs to me would be to have a file somewhere with the versions in it that Cursive could download at startup or something, but that would have to be produced as part of the gradle-clojure build. Or I could add some functionality to search Maven repos for the latest version of a particular artifact, and always use the latest? I'm not sure what the right thing to do here is.

**Comment by ajoberstar** [comment]
Thanks Colin! Looks good so far.

It might not be as simple as just picking up the latest versions, at least in the early days. There could be some breaking changes to what we would want as a default project.

Is it possible to have the source of those sample projects (or the code that generates them) somewhere the community could contribute to? (Obviously, Cursive itself is closed source and I don't mean to compromise that, so not sure how this would work.)

**Comment by cursive-ide** [comment]
The issue is that this is all generated by the IntelliJ Gradle support, and it's surprisingly dumb:

```java
    buildScriptData
        .addPluginDefinitionInPluginsGroup("id 'gradle-clojure.clojure' version '0.3.1'")
        .addRepositoriesDefinition("jcenter()")
        .addRepositoriesDefinition("maven {\n" +
                                   "        name = 'clojars'\n" +
                                   "        url = 'https://clojars.org/repo/'\n" +
                                   "    }")
        .addDependencyNotation("compile 'org.clojure:clojure:1.9.0'")
        .addDependencyNotation("testCompile group: 'junit', name: 'junit', version: '4.12'")
        .addDependencyNotation("devCompile 'org.clojure:tools.namespace:0.3.0-alpha4'")
```

So basically I just return a set of strings, and they get inserted into the result. If for example I used `testCompile 'junit:junit:4.12'` then anyone selecting both Clojure and Java would get the JUnit dep twice, once in each format, because it just uses a simple string compare.

**Comment by cursive-ide** [comment]
Oh, re-reading your comment, if you meant that those details could be provided somewhere that would be bundled with Cursive, we could definitely do that. But the problem is that the versions will be hard-coded to a particular version of Cursive, which isn't ideal (i.e. Cursive 1.7 will always create projects with gradle-clojure 0.3.1).

**Comment by ajoberstar** [comment]
Yeah, that is a pretty unfortunate API.

Let me confirm I'm following the options:

1. Cursive could just output static build scripts and make updates in new Cursive versions, as needed.
1. Cursive could dynamically look up versions of the plugins/dependencies to include in the build scripts, but if the general structure of the build script changed, this wouldn't be handled.
1. Cursive could bundle some external JAR (?) that included the code (like you provided a sample of above). This could potentially be open source so community could contribute updates. However, a particular Cursive version would always output the same builds. So a change to that external JAR would require a Cursive update to be published to have it included.

If I have that right, I feel like option 1 or 3 are the way to go. I'd rather have a working, but outdated, build produced than a potentially invalid one with the dynamic versions. As kind of a side question, how up-to-date do your customers tend to be with the Cursive plugin (if you have that info)?

Option 3 would make it easiest for me and the rest of the community to provide any updates, but since that's an impact on Cursive, I'm cool with whatever works for you.

**Comment by cursive-ide** [comment]
I think that's a fair summary. For reference, [here](https://github.com/JetBrains/intellij-community/blob/master/plugins/gradle/src/org/jetbrains/plugins/gradle/frameworkSupport/BuildScriptDataBuilder.java) is the class used to build the build script file. As you can see, the options to customise the build are pretty limited. Option 2 could be extended by downloading an EDN file which contains a map whose values are a vector of strings to be added to the different build script sections. Something like:

```clojure
{:plugin-group-plugins ["id 'gradle-clojure.clojure' version '0.3.1'"']
 :repositories ["jcenter()"
                "maven {
        name = 'clojars'
        url = 'https://clojars.org/repo/'
    }"]
 :dependencies ["compile 'org.clojure:clojure:1.9.0'"
                "testCompile group: 'junit', name: 'junit', version: '4.12'"
                "devCompile 'org.clojure:tools.namespace:0.3.0-alpha4'"]}
```

This could support a `:buildscript` sub-map containing elements to be added to the buildscript block, and as long as it handles entries for `:imports`, `:properties` and `:other` that would support all the customisation that that API supports anyway. Then gradle-clojure could publish the EDN file containing the current recommended settings somewhere, and Cursive could bundle the most recent version when each build is cut. Then Cursive could try to dynamically download the most recent version, and fall back on the bundled one if it can't for whatever reason.

**Comment by ajoberstar** [comment]
That EDN approach seems pretty promising. How would the testing process for that work? I'd be worried about publishing an update before verifying it.

**Comment by cursive-ide** [comment]
I haven't thought about the testing for this yet, but I have found another problem in the meantime. The `addPluginDefinitionInPluginsGroup` was only added in 2017.3. So I think that for earlier versions, I'll have to generate code which adds an explicit dependency to the `buildscript` block, and then `apply`s the plugin. I'm not sure exactly what that should look like, do you @ajoberstar? Something like this?:
```
buildscript {
  dependencies {
    compile 'io.github.gradle-clojure:gradle-clojure-plugin:0.4.0-beta.9'
  }
}

apply plugin: 'gradle-clojure.clojure'
```

**Comment by ajoberstar** [comment]
@cursive-ide Pretty close:

```groovy
 buildscript {
  repositories {
    gradlePluginPortal()
  }
  dependencies {
    classpath 'io.github.gradle-clojure:gradle-clojure-plugin:0.4.0-beta.9'
  }
}

apply plugin: 'gradle-clojure.clojure'
```

Main piece is `classpath` instead of `compile`. It might have `gradlePluginPortal()` in there by default. I know it does for the `plugins {}` block.

**Comment by cursive-ide** [comment]
Thanks. One question - do you think it's actually worth making this change, and generating projects using an obsolete plugin inclusion mechanism? Since JetBrains moved to subscription, most people are on recent versions - perhaps it would be better to just not have the Clojure option on older IntelliJ versions? Obviously users on those versions can modify the standard build files themselves following the doc.

**Comment by ajoberstar** [comment]
@cursive-ide Seems fine to me that it's only on newer versions of IntelliJ. I think the clj-new templates will serve as a good backup for someone on an older version.

**Comment by neuromantik33** [comment]
Hi there is an issue I'm encountering with clojurephant and cursive. Well it works for the most part for a simple gradle project, but in order for intellij to recognize the source roots I couldn't use the clojure dir in the main sourceset but I needed to put all clj files in the java dir. Same things go for dev and test sourcesets. Can anyone confirm this? I can provide screenshots if needed.

**Comment by neuromantik33** [comment]
Here is a screenshot of what I'm experiencing with cursive

![clojurephant](https://user-images.githubusercontent.com/19826274/75037906-7f503c80-54b5-11ea-8d1b-f9fe9645aac0.png)

It is unfortunately a deal breaker for me if I cannot properly work with it in the IDE of choice. Any suggestions on workarounds besides the one I mentioned which is to apply the following changes to my build:
```groovy
sourceSets {
    main {
        java { srcDirs = ['src'] }
        resources { srcDirs = ['resources'] }
    }
    dev {
        java { srcDirs = ['dev'] }
        resources { srcDirs = ['dev-resources'] }
    }
    test {
        java { srcDirs = ['test'] }
        resources { srcDirs = ['test-resources'] }
    }
}
```

**Comment by chris-pollard-assurity** [comment]
@neuromantik33, I'm working through something similar in IntelliJ+Cursive.

By way of an alternative work around one trick I've used before to get IntelliJ to see additional source modules or paths is with the `idea` gradle plugin and some lines like:
```
idea {
    module {
        sourceDirs += project.sourceSets.main.clojure.srcDirs
        testSourceDirs += project.sourceSets.test.clojure.srcDirs
    }
}
```
This is working, if I use the source and test paths as follows in my project:
```
//src/main/clojure
//src/test/clojure
```
then IDEA recognises these directories as module Source Folders and Test Source Folders 
I am still working on getting the `dev` module to work though...

IDEA seems to be doing the right thing - myname.myapp-test fails appropriately (or passes once fixed) when run by IDEA and I can config an nREPL, load the myapp namespace and run (-main)

**Comment by neuromantik33** [comment]
> @neuromantik33, I'm working through something similar in IntelliJ+Cursive.
> 
> By way of an alternative work around one trick I've used before to get IntelliJ to see additional source modules or paths is with the `idea` gradle plugin and some lines like:
> 
> ```
> idea {
>     module {
>         sourceDirs += project.sourceSets.main.clojure.srcDirs
>         testSourceDirs += project.sourceSets.test.clojure.srcDirs
>     }
> }
> ```
I was under the impression that the gradle `idea` plugin was deprecated but I'll take another look at it. Thanks for the tip ;)

**Comment by pbzdyl** [comment]
Hi @cmf, @ajoberstar. Do you have any suspects where the issue occurs with the source/test folders not being recognized properly? Would it be clojurephant (e.g. by not annotating properly some of the gradle build model objects like source sets) or rather in Cursive (where it doesn't report these directories to Intellij) or something else?

**Comment by ajoberstar** [comment]
@pbzdyl Sorry, I don't know much about this issue. Haven't looked at it in a long time.

**Comment by ajoberstar** [comment]
@pbzdyl I think @neuromantik33 might be on the right track, however I'm not positive that IntelliJ leverages the `idea` plugin when you do an import. I'm just not terribly familiar with Intellij.

But see this part of the [IdeaPlugin](https://github.com/gradle/gradle/blob/d9f5f8e83497eb3dd4c50d3d9d0eaf7e773f84cf/subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java#L369) where they are explicitly registering the source directories and reference the `getAllJava` method. From memory, I think languages like Groovy add their source into the `allJava` part of the sourceSet. I'm not positive if that's right for Clojure, but could at least be a pointer in the right direciton.

**Comment by cursive-ide** [comment]
Unfortunately I don't know anything about this. Cursive plays no part in this, it's the Gradle integration which does it.

**Comment by ajoberstar** [comment]
Source root detection should be fixed by #177. If there's still interest in other features, let's open new issues.
